docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,341 @@
+# ADR-102: RPC Companion
+
+## Changelog
+
+- 2022-03-27: First draft (@andynog)
+
+## Status
+
+Accepted | Rejected | Deprecated | Superseded by
+
+## Context
+
+This ADR proposes an architecture of an ***RPC Companion*** solution implemented based on the proposed [ADR-101 Data Companion Pull API](https://github.com/cometbft/cometbft/blob/thane/adr-084-data-companion-pull-api/docs/architecture/adr-101-data-companion-pull-api.md). This solution can run as a sidecar concurrently with the full node, and it is optional.
+
+This ADR provides a reference implementation of a system that can be used to offload queryable data from a CometBFT
+full node to a database and offering a service that exposes the same JSONRPC methods on a endpoint as the regular JSONRPC methods of a CometBFT node endpoint,
+which makes it easier for integrators of RPC clients such as client libraries and applications to switch to this
+***RPC Companion*** with as minimum effort as possible.
+
+This architecture also makes it possible to scale horizontally the querying capacity of a full node by running multiple
+copies of the ***RPC Companion*** server instances that can be behind a scalable load-balancer (e.g. Cloudflare)  and can serve
+the data in a more reliable way.
+
+## Alternative Approaches
+
+Currently, there aren't any alternative solutions that are based on a data companion architecture. It is
+expected that integrators and operators propose and design their own solutions based on their specific use-cases.
+
+This ADR provides a reference implementation that can be used and adapted for individual use-cases.
+
+## Decision
+
+TBD
+
+## Detailed Design
+
+### Requirements
+
+The target audience for this solution are operators and integrators that want to alleviate the load on their nodes by offloading
+the queryable data requests to the **RPC Companion**.
+
+This solution shall meet the following requirements in order to provide real benefits to these users.
+
+The **RPC Companion** solution shall:
+
+1. Provide an **[Ingest Service](#ingest-service)** implemented as a data companion that can pull data from the node and store it on
+its own storage (database)
+2. Provide its own storage ([Database](#database)) that can handle a high-load of reads and also have a good performance on inserting data.
+3. Implement a [database schema](#database-schema) that needs to be backwards compatible with the current CometBFT RPC.
+4. Do not enforce any breaking changes to the existing RPC.
+5. Ensure the responses returned by the [RPC Companion v1 endpoint](#rpc-endpoint) is wire compatible with the existing CometBFT endpoint.
+6. Implement tests to verify backwards compatibility.
+7. Be able to handle multiple concurrent requests and return idempotent responses.
+8. Do not crash or panic if the querying demand is very high or large responses are returned.
```

Comment:
This requirement is... odd.

Reply:
Yes, I suspect we should remove this requirement. Not sure it's implementable.

Reply:
makes sense, removed